The purpose of this notebook is to understand the basics of neural networks.

We will start with a single neuron and then move on to a simple neural network.





In [2]:
import torch

torch.manual_seed(0)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(1, 1)

    def forward(self, x):
        # Pass through the linear layer (single neuron)
        return self.linear(x)


# Create an instance of the model
model = Model()
print(f"{model=}")
# Now call the model instance with the input
print(f"{model(torch.tensor([0.0]))=}")


model=Model(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
model(torch.tensor([0.0]))=tensor([0.5364], grad_fn=<ViewBackward0>)


The output of the model for the input of 0.0 is a tensor with a single value that is the result of a random initialization of the weights and biases.

By default, autograd is tracking the operations that are performed on the model.
`grad_fn=<ViewBackward0>` tells us that the last operation to create this output was a view operation. Without the view operation the output would have been [[0.5364]] rather than [0.5364]. The `Backward` means that during backgpropagation the Backward function will be called. The `0` suffix means that this will be the first backward call.







In [ ]:
# Let's create a simple training example

# input
x = torch.tensor([[0.0]])
# target output
y = torch.tensor([[1.0]])

criterion = torch.nn.MSELoss()

# Forward pass
output = model(x)
loss = criterion(output, y)
print(f"Initial prediction: {output.item()}")
print(f"Loss: {loss.item()}")

# Backward pass
loss.backward()  # This computes gradients for all model parameters

# Print gradients
for name, param in model.named_parameters():
    print(f"Parameter: {name}, Gradient: {param.grad}")


Initial prediction: 0.5364435911178589
Loss: 0.21488454937934875
Parameter: linear.weight, Gradient: tensor([[0.]])
Parameter: linear.bias, Gradient: tensor([-0.9271])
